### ISSUE TITLE
UDP forwarding doesn't seem to work #2

### ISSUE THREAD

**Issue by Kein** [shadowsocks/shadowsocks-android#2423]
**Describe the bug**
UDP/DNS relay does not work properly

**To Reproduce**
1. Enable DNS/UDP relay in client and UDP support on server
2. Go to https://www.dnsleaktest.com/
3. Run extended test
4. See results dont quite match

**Expected behavior**
DNS server detected by test should make sense

**Screenshots**

**Smartphone (please complete the following information):**
 - Android/Chrome OS version: 9PKQ1.180904.001
 - Device: Redmi Note 7
 - Version: 4.8.7a

**Configuration**
* [x] IPv4 server address
* [ ] IPv6 server address
* [x] Client IPv4 availability
* [ ] Client IPv6 availability
* Encrypt method: X-CHACHA20-IETF-POLY305
* Route
  * [ ] All
  * [x] Bypass LAN
  * [ ] Bypass China
  * [ ] Bypass LAN & China
  * [ ] GFW List
  * [ ] China List
  * [ ] Custom rules
* [ ] IPv6 route
* [ ] Apps VPN mode
  * [ ] Bypass mode
* Remote DNS: default (1dot1dot1dot1dot.cloudflare-dns.com)
* [x] DNS over UDP
* Plugin configuration (if applicable): none
* [ ] Auto Connect
* [x] TCP Fast Open

**Additional context**
After a numerous tests I came to the same conclusion as closed now #1090 and I think the issue does exist. Here is a quick example with the same proxy (let's call it `ProxyMain`) on desktop and on Android:

**Dekstop:**
Firefox is setup to use ProxyMain shadowsock proxy and "Send DNS over proxy" is enabled. Result at https://dnsleaktest.com/ - proyx's local DNS is used:
![image](https://user-images.githubusercontent.com/623756/73288156-87e18a00-420b-11ea-8858-639ce5ebfa94.png)

**Android:**
Result from DNSLeak - All dns' servers are Cloudflare for the EXACT same ProxyMain shadowsocks proxy. Used ungoogle Chromium, DoH diabled, local dns - default google's 8.8.8.8

**Comment by Kein** [comment]
Upon some more investigation it would appear that you have completely changed behavior of  `nameserver` as opposed to other implementations. It was supposed to be an internal DNS SS uses when it needs to resolve a hotsname firsthand, like, for example, if you specify your proxy address as a hostname and connection obviously not established yet. But in case of SS-android it seems like it is acting like THE DNS server, replacing everything else. In this case the whole UDP/DNS relay sure makes little sense. User expects it to relay the DNS (udp) queries directly like you would get with VPN or tunnelled (via virtual TAP/TUN ) Shadowsocks but instead the all queries are intercepted and internal DNS used?
That is besides the fact there is still leak somewhere which I can reproduce randomly.

**Comment by Mygod** [comment]
This is working as expected. To prevent DNS leaks, use route: all. If you choose bypass LAN, DNS requests will be sent to local DNS server and the results will be used if the address returned is a LAN address.

**Comment by Mygod** [comment]
Fixed possible inconsistency that might be causing the confusion via 1e0ae01.

**Comment by Kein** [comment]
> If you choose bypass LAN, DNS requests will be sent to local DNS server and the results will be used if the address returned is a LAN address.

Could you rephrase it? I read it 13 times and still failed to grasp exact intricacies. What address will be used? Localhost? Local DNS specified in props/resolv? Shadowsock remote DNS? What happens if I disable bypass completely?
